oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryImpl.java

```diff
@@ -625,19 +625,15 @@ private boolean canSortByIndex() {
                         OrderEntry e = list.get(i);
                         OrderingImpl o = orderings[i];
                         DynamicOperandImpl op = o.getOperand();
-                        if (!(op instanceof PropertyValueImpl)) {
-                            // ordered by a function: currently not supported
-                            canSortByIndex = false;
-                            break;
-                        }
                         // we only have one selector, so no need to check that
                         // TODO support joins
-                        String pn = ((PropertyValueImpl) op).getPropertyName();
+                        String pn = op.getOrderEntryPropertyName(selectors.get(0));
```

Comment:
getOrderEntryPropertyName can in theory return null, then you get a NullPointerException in QueryImpl. I know, it can only happen in theory, but avoiding a potential NPE would be good. I think it should be relatively easy, for example throw a different exception with an appropriate message, if getOrderEntryPropertyName returns null.

Reply:
@thomasmueller  - Instead of throwing an exception - maybe we can simply set canSortByIndex = false in QueryImpl in case getOrderEntryPropertyName returns null (that way the query would still return result , sorted by QueryEngine). WDYT ?